tests/app/authentication/test_database_session_manager.py

```diff
@@ -1,49 +1,124 @@
 import unittest
 
+import app
+from app.data_model.database import EQSession
+from app.authentication.session_storage import SessionStorage, EQ_SESSION_ID
+
+import flask
 from mock import patch, Mock
 from sqlalchemy.exc import IntegrityError
 
-from app.authentication.session_storage import SessionStorage, EQ_SESSION_ID
-
+# SQLAlchemy confuses pylint and we want to override a protected method here
+# pylint: disable=no-member,protected-access
 
 class TestSessionManager(unittest.TestCase):
 
-    @staticmethod
-    def test_store_user_id_rollback():
-        # Given
-        session_storage = SessionStorage()
-        user_id = "1"
+    def setUp(self):
+        self.flask_app = app.create_app()
+        self.session_storage = SessionStorage()
+
+        # Create a patched db_session so we don't need a real database to test against
+        self.db_session_patcher = patch('app.authentication.session_storage.db_session', autospec=True)
+        self.addCleanup(self.db_session_patcher.stop)
+        self.db_session = self.db_session_patcher.start()
+
+        # Note each test will need to create a test_request_context
+        # in order to access the Flask session object
+
+    def test_store_new_user_id_then_replace(self):
+        with self.flask_app.test_request_context('/status'):
+            # Initially there is no session id present
+            self.assertFalse(EQ_SESSION_ID in flask.session)
+
+            # Store a user id for the first time
+            self.session_storage.store_user_id('1')
+
+            # When we store the user_id we generate a GUID and use that
+            # as the session_id mapped to the user_id in the database
+            self.assertTrue(EQ_SESSION_ID in flask.session)
+            self.assertIsNotNone(flask.session[EQ_SESSION_ID])
+            self.assertTrue(self.db_session.add.called)
+            self.assertEqual(self.db_session.add.call_count, 1)
+
+            # Store the GUID so we can check it changes
+            eq_session_id = flask.session[EQ_SESSION_ID]
+
+            # Try replacing the session with a new one
+            self.session_storage.store_user_id('2')
+
+            # Session GUID should have changed
+            self.assertEqual(self.db_session.add.call_count, 2)
+            self.assertNotEqual(eq_session_id, flask.session[EQ_SESSION_ID])
+
+    def test_clear(self):
```

Comment:
This test seems over complicated and should probably be 3 tests. Each test should fail for it's own reason. This makes it easier to diagnose and fix the problem.
- `test_should_not_clear_with_no_session_data`
- `test_should_not_clear_with_no_session_data`
- `test_should_clear_with_session_and_data`